### PR TITLE
Branch Content Discovery 

### PR DESCRIPTION
**Pull request by sojanpr**
Branch Content Discovery implementation for next step content analytics.

@aaustin @ahmednawar

### CODE REVIEW

File: Branch-SDK/Branch-SDK/BNCConfig.h [hunk 1]
```diff
@@ -16,7 +16,7 @@
 //#define BNC_DEV_ENV
 
 #ifdef BNC_PROD_ENV
-#define BNC_API_BASE_URL        @"https://api.branch.io"
+#define BNC_API_BASE_URL        @"http://chandra.branch.io:8080" //@"https://api.branch.io"
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/BNCConfig.h]: make sure we change that before releasing :)

File: Branch-SDK/Branch-SDK/BNCPreferenceHelper.m [hunk 2]
```diff
@@ -56,29 +58,30 @@ @interface BNCPreferenceHelper ()
...
-            shouldWaitForInit = _shouldWaitForInit,
-            suppressWarningLogs = _suppressWarningLogs,
-            retryCount = _retryCount,
-            retryInterval = _retryInterval,
-            timeout = _timeout,
-            lastStrongMatchDate = _lastStrongMatchDate,
-            checkedFacebookAppLinks = _checkedFacebookAppLinks,
-            requestMetadataDictionary = _requestMetadataDictionary,
-            instrumentationDictionary = _instrumentationDictionary;
+lastRunBranchKey = _lastRunBranchKey,
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/BNCPreferenceHelper.m]: can we revert the indentation here?

File: Branch-SDK/Branch-SDK/Branch.m [hunk 3]
```diff
@@ -33,6 +33,7 @@
 #import "BranchInstallRequest.h"
 #import "BranchSpotlightUrlRequest.h"
 #import "BranchRegisterViewRequest.h"
+#import "ContentDiscoverer.h"
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/Branch.m]: nitpick: can we call it `BranchContentDiscoverer` instead?

File: Branch-SDK/Branch-SDK/ContentDiscoverer.m [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+- (void)readContentData {
+    UIViewController *viewController = _lastViewController;
+    if (viewController) {
+        UIView *rootView = [viewController view];
+        if ([viewController isKindOfClass:UITableViewController.class]) {
+            rootView = ((UITableViewController *)viewController).tableView;
+        } else if ([viewController isKindOfClass:UICollectionViewController.class]) {
+            rootView = ((UICollectionViewController *)viewController).collectionView;
+        }
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/ContentDiscoverer.m]: what about tabbed view controllers?

File: Branch-SDK/Branch-SDK/ContentDiscoverer.m [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+    } else {
+        [self stopContentDiscoveryTask];
+    }
+}
+
+- (void)readContentData {
+    UIViewController *viewController = _lastViewController;
+    if (viewController) {
+        UIView *rootView = [viewController view];
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/ContentDiscoverer.m]: nitpick: abstract this to a method

File: Branch-SDK/Branch-SDK/ContentDiscoverer.m [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+    if (contentData && contentData.length > _cdManifest.maxTextLen) {
+        contentData = [contentData substringToIndex:_cdManifest.maxTextLen];
+    }
+    if (!isClearText) {
+        contentData = [self hashContent:contentData];
+    }
+    [contentDataArray addObject:contentData];
+}
+
+- (NSString*)hashContent:(NSString *)content {
```
Comment (ahmednawar) [inline Branch-SDK/Branch-SDK/ContentDiscoverer.m]: space before the `*` in the return type
might be a good candidate to add to the encoding utils class

**Comment by ahmednawar** [comment]
@sojanpr I still don't understand all the nuances but looks good. I would say let's test as much as possible to make sure we are covering most cases with the ViewController stuff

**Comment by ahmednawar** [comment]
@sojanpr also would be great if you re-open the PR against `staging` instead of `master` if possible.

File: Branch-SDK/Branch-SDK/BNCPreferenceHelper.m [hunk 7]
```diff
@@ -542,6 +540,52 @@ - (NSInteger)getBranchViewCount:(NSString *)branchViewID {
...
+
+- (NSMutableDictionary *)getBranchAnalyticsData {
+    NSMutableDictionary *analyticsDataObj = _savedAnalyticsData;
+    if (!analyticsDataObj) {
+        NSObject *dataObj = [self readObjectFromDefaults:KEY_BRANCH_ANALYTICAL_DATA];
+        if (dataObj) {
+            analyticsDataObj = (NSMutableDictionary *)dataObj;
+        } else {
+            analyticsDataObj = [[NSMutableDictionary alloc] init];
+        }
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCPreferenceHelper.m]: At the risk of sounding to nitpickey, but you asked for it:
```
analyticsDataObj = (NSMutableDictionary *)[self readObjectFromDefaults:KEY_BRANCH_ANALYTICAL_DATA];
if (!dataObj) {
    analyticsDataObj = [[NSMutableDictionary alloc] init];
}
```

Comment (sojanpr) [inline Branch-SDK/Branch-SDK/BNCPreferenceHelper.m]: Good catch!
Was scared of any potential issue with  typecasting a  `nil`. Just learned it is not an issue on iOS

File: Branch-SDK/Branch-SDK/BNCPreferenceHelper.m [hunk 8]
```diff
@@ -542,6 +540,50 @@ - (NSInteger)getBranchViewCount:(NSString *)branchViewID {
...
+    }
+    return analyticsDataObj;
+}
+
+- (void)saveContentAnalyticsManifest:(NSDictionary *)cdManifest {
+    [self writeObjectToDefaults:KEY_BRANCH_ANALYTICS_MANIFEST value:cdManifest];
+}
+
+- (NSDictionary *)getContentAnalyticsManifest {
+    NSObject *manifestObj = [self readObjectFromDefaults:KEY_BRANCH_ANALYTICS_MANIFEST];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCPreferenceHelper.m]: this code requires the same cleanup. Here you go:
```
return (NSDictionary *)[self readObjectFromDefaults:KEY_BRANCH_ANALYTICS_MANIFEST];
```

Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCPreferenceHelper.m]: Can you go through and find all of these patterns and clean it up?

File: Branch-SDK/Branch-SDK/BNCPreferenceHelper.m [hunk 9]
```diff
@@ -39,6 +39,8 @@
 NSString * const BRANCH_PREFS_KEY_CREDIT_BASE = @"bnc_credit_base_";
 
 NSString * const BRANCH_PREFS_KEY_BRANCH_VIEW_USAGE_CNT = @"bnc_branch_view_usage_cnt_";
+NSString * const KEY_BRANCH_ANALYTICAL_DATA = @"bnc_branch_analytical_data";
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCPreferenceHelper.m]: can you follow the same naming style right above this? `BRANCH_PREFS_...`

File: Branch-SDK/Branch-SDK/BranchContentDiscoverer.m [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+                }
+                
+                [contentEventObj setObject:[NSString stringWithFormat:@"/%@", _lastViewControllerName] forKey:BRANCH_VIEW_KEY];
+                [contentEventObj setObject:!isClearText? @"true" : @"false" forKey:BRANCH_HASH_MODE_KEY];
+                [contentEventObj setObject:contentKeysArray forKey:BRANCH_CONTENT_KEYS_KEY];
+                if (contentDataArray && contentDataArray.count > 0) {
+                    [contentEventObj setObject:contentDataArray forKey:BRANCH_CONTENT_DATA_KEY];
+                }
+                
+                [[BNCPreferenceHelper preferenceHelper]saveBranchAnalyticsData:contentEventObj];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchContentDiscoverer.m]: space here

File: Branch-SDK/Branch-SDK/BranchContentDiscoverer.m [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+            if (contentVal) {
+                viewTxt = contentVal;
+            }
+        }
+    }
+    return viewTxt;
+}
+
+- (UIViewController *)getActiveViewController {
+    UIViewController *rootViewController = [UIApplication sharedApplication].keyWindow.rootViewController;
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchContentDiscoverer.m]: Can you use `Class UIApplicationClass = NSClassFromString(@"UIApplication");`, then replace this line with `UIViewController *rootViewController = [UIApplicationClass sharedApplication].keyWindow.rootViewController;` We can't call UIApplication in iMessage extension. Also, sorry to add to the scope, but can you make sure the calling methods handle `nil` properly in this case?

File: Branch-SDK/Branch-SDK/BranchContentDiscoverer.m [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        activeController = ((UINavigationController *)rootViewController).topViewController;
+    } else if ([rootViewController isKindOfClass:[UITabBarController class]]) {
+        activeController = ((UITabBarController *)rootViewController).selectedViewController;
+    } else {
+        activeController = rootViewController;
+    }
+    return activeController;
+}
+
+- (void)addFormatedContentData:(NSMutableArray *)contentDataArray withText:(NSString *)contentData clearText:(BOOL)isClearText {
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchContentDiscoverer.m]: Add another `t` to Formatted.

File: Branch-SDK/Branch-SDK/BranchContentDiscoverer.m [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    if (contentData && contentData.length > _cdManifest.maxTextLen) {
+        contentData = [contentData substringToIndex:_cdManifest.maxTextLen];
+    }
+    if (!isClearText) {
+        contentData = [self hashContent:contentData];
+    }
+    [contentDataArray addObject:contentData];
+}
+
+- (NSString *)hashContent:(NSString *)content {
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchContentDiscoverer.m]: Can you use this method instead? https://github.com/BranchMetrics/ios-branch-deep-linking/blob/master/Branch-SDK/Branch-SDK/BNCEncodingUtils.h#L17 It's battle hardened.

File: Branch-SDK/Branch-SDK/BranchContentDiscoveryManifest.m [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+@end
+
+@implementation BranchContentDiscoveryManifest
+
+
+static BranchContentDiscoveryManifest *contentDiscoveryManifest;
+
+- (instancetype)init
+{
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchContentDiscoveryManifest.m]: small nit: put bracket on previous line

File: Branch-SDK/Branch-SDK/BranchContentPathProperties.m [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    return self;
+}
+
+- (NSArray *)getFilteredElements {
+    return [_pathInfo objectForKey:BRANCH_FILTERED_KEYS];
+}
+
+- (BOOL)isSkipContentDiscovery {
+    NSArray *filteredElelments = [self getFilteredElements];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BranchContentPathProperties.m]: elements typo

**Comment by aaustin** [comment]
Outside of the small nits, I think this is ready. Can you pull against staging?

**Pull request closed without merging**